perfkitbenchmarker/scripts/object_storage_api_tests.py

```diff
@@ -187,32 +187,55 @@ class SizeDistributionIterator(object):
     dist: dict. The distribution to draw sizes from.
   """
 
-  # IMPLEMENTATION LIMITATION: right now, this object only supports
-  # point distributions, which contain only a single size.
-
   def __init__(self, dist):
-    if len(dist) != 1:
-      raise NotImplementedError(
-          'General distributions are not implemented yet.')
-
-    size, _ = dist.iteritems().next()
-    self.size = size
+    # This class chooses random numbers by using random.random() to
+    # pick a float in [0.0, 100.0), then converting that into a choice
+    # from our distribution. The data structure is two lists, sizes
+    # and cutoffs. If the random number is in the range [cutoffs[i-1],
+    # cutoffs[i]), then the random choice is sizes[i]. Cutoffs has an
+    # implied starting element 0.0, and its last element must be
+    # 100.0.
+    self.sizes = []
+    self.cutoffs = []
+
+    total_percent = 0.0
+    for size, percent in dist.iteritems():
+      total_percent += percent
+      self.sizes.append(size)
+      self.cutoffs.append(total_percent)
+
+    # I considered storing probabilities instead of percents, but I'm
+    # afraid that there is an edge case where the probability
+    # corresponding to a percent doesn't have a terminating binary
+    # representation and then the total percent is less than 1.0 even
+    # though the percentages add up to 100. I can't prove that this
+    # happens, but storing percentages seems safer.
+    assert total_percent == 100.0
 
   # this is required by the Python iterator protocol
   def __iter__(self):
     return self
 
   def next(self):
+    val = random.random() * 100.0
+
+    # binary search has a lower asymptotic complexity than linear
+    # scanning, but thanks to caches and sequential access, I suspect
+    # linear scanning will be faster for any realistic distribution.
+    for i in xrange(len(self.sizes)):
+      if self.cutoffs[i] > val:
+        return self.sizes[i]
+
+    raise AssertionError("Random percent %s didn't match percent cutoff list %s"
+                         % (val, self.cutoffs))
+
     return self.size
 
 
 def MaxSizeInDistribution(dist):
   """Find the maximum object size in a distribution."""
 
-  if len(dist) != 1:
-    raise NotImplementedError('General distributions are not implemented yet.')
-
-  return long(dist.iterkeys().next())
+  return max(dist.iterkeys())
```

Comment:
I'm not sure which initialization function you mean. The API test script initialization? The `SizeDistributionIterator` initialization?

If you mean the `SizeDistributionIterator` initialization, that object itself won't work well because we construct the iterators in the worker threads, but `MaxSizeInDistribution` is called in the parent thread to construct the payload. I think the cleanest way to handle this would be to make an object that represents a distribution. It would have methods to get an iterator and to find the maximum object size in the distribution. I'm just not sure it's worth the extra code it would take to implement that. What would you like me to do?